### PR TITLE
Make it so that MixedRealityRaycastHit doesn't throw an exception every frame.

### PR DESCRIPTION
**Pull request by wiwei**
https://github.com/microsoft/MixedRealityToolkit-Unity/issues/4966

When the default raycaster goes through a physics hit and doesn't hit anything, the generated RaycastHit isn't usable. The raycaster returns true when a hit is successful (the only time when the out parameter should be accessed) but otherwise the out parameter shouldn't be used (and it should be null instead)

 lightmapCoord itself is kinda funky when the physics hit doesn't hit anything (if you see the stuff here https://github.com/Unity-Technologies/UnityCsReference/blob/master/Modules/Physics/ScriptBindings/Dynamics.bindings.cs you'll find that it actually tries to do stuff, but the object basically isn't valid to do stuff.) Since RaycastHit is a struct it's not a nullable type, but those objects really aren't safe to use in the case where the physics hit failed.

Also adds a test to show how this all works end to end (that it doesn't throw, and that it's null in that case).

### CODE REVIEW

**Comment by wiwei** [comment]
Test failure I think is related to the way I didn't comment out the tests for UWP, will do that shortly.

File: Assets/MixedRealityToolkit/Interfaces/InputSystem/IMixedRealityRaycastProvider.cs [hunk 1]
```diff
@@ -19,12 +19,18 @@ public interface IMixedRealityRaycastProvider : IMixedRealityDataProvider
         /// <summary>
         /// Performs a raycast using the specified <see cref="Microsoft.MixedReality.Toolkit.Physics.RayStep"/>.
         /// </summary>
+        /// <remarks>
+        /// If nothing was hit, hitInfo will be null.
```
Comment (Railboy) [inline Assets/MixedRealityToolkit/Interfaces/InputSystem/IMixedRealityRaycastProvider.cs]: We're already using the bool / out pattern that Unity uses for raycast hits. Shouldn't we just return false if we hit nothing, instead of making hitInfo null?

File: Assets/MixedRealityToolkit/Definitions/InputSystem/MixedRealityRaycastHit.cs [hunk 2]
```diff
@@ -9,7 +9,7 @@ namespace Microsoft.MixedReality.Toolkit.Input
     /// <summary>
     /// The resulting hit information from an IMixedRealityRaycastProvider.
     /// </summary>
-    public struct MixedRealityRaycastHit
+    public class MixedRealityRaycastHit
```
Comment (Railboy) [inline Assets/MixedRealityToolkit/Definitions/InputSystem/MixedRealityRaycastHit.cs]: We were using a struct so that people could use MixedRealityRaycastHit in the same way as RaycastHit, and to make avoiding garbage easier.

If it's necessary to tell when a MixedRealityRaycastHit is 'valid' we could add a HitSomething field or a read-only Empty property that returns true if certain values haven't been set. Or we could just rely on the bool / out pattern that's already in place.

Comment (jbienzms) [reply Assets/MixedRealityToolkit/Definitions/InputSystem/MixedRealityRaycastHit.cs]: Actually, @Railboy makes a very good point here @wiwei. These raycasts happen _very_ frequently and on the UI thread. If a Class is used instead of a Struct it's going to add _significant_ pressure on the garbage collector. If there is any way to use a struct instead here, it would be highly desirable.

**Comment by jbienzms** [comment]
Awesome work. Thanks @wiwei !
